extensions/amp-a4a/0.1/test/test-refresh-manager.js

```diff
@@ -0,0 +1,117 @@
+} from '../refresh-manager';
+import * as sinon from 'sinon';
+
+function getNumElementsRegistered(refreshManager) {
+  return Object.keys(refreshManager.registeredElementWrappers_).length;
+}
+
+function getTestElement() {
+  const div = window.document.createElement('div');
+  div.setAttribute('style', 'width:1px; height:1px;');
+  return div;
+}
+
+
+describe('refresh-manager', () => {
+  let refreshManager;
+  let testElement;
+
+  const noop = () => {};
+
+  beforeEach(() => {
+    refreshManager = getRefreshManagerFor(window);
+    testElement = getTestElement();
+  });
+
+  afterEach(() => {
+    refreshManager = null;
+    testElement = null;
+    resetRefreshManagerFor(window);
+  });
+
+  describe('#registerElement', () => {
+
+    it('should have no registered elements', () => {
+      expect(getNumElementsRegistered(refreshManager)).to.equal(0);
+    });
+
+    it('should have one registered element', () => {
+      expect(getNumElementsRegistered(refreshManager)).to.equal(0);
+      refreshManager.registerElement(testElement, noop);
+      expect(getNumElementsRegistered(refreshManager)).to.equal(1);
+    });
+
+    it('should add correct attribute to element', () => {
+      refreshManager.registerElement(testElement, noop);
+      expect(testElement.getAttribute(REFRESH_REFERENCE_ATTRIBUTE))
+          .to.equal('0');
+    });
+
+    it('should have multiple registered elements', () => {
+      expect(getNumElementsRegistered(refreshManager)).to.equal(0);
+      refreshManager.registerElement(testElement, noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      expect(getNumElementsRegistered(refreshManager)).to.equal(5);
+    });
+  });
+
+  describe('#resetManager', () => {
+    it('should reset manager to initial state', () => {
+      expect(getNumElementsRegistered(refreshManager)).to.equal(0);
+      refreshManager.registerElement(testElement, noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      refreshManager.registerElement(getTestElement(), noop);
+      expect(getNumElementsRegistered(refreshManager)).to.equal(5);
+      expect(refreshManager.elementReferenceId_).to.equal(5);
+      refreshManager.resetManager();
+      expect(getNumElementsRegistered(refreshManager)).to.equal(0);
+      expect(refreshManager.elementReferenceId_).to.equal(0);
+    });
+  });
+
+  describe('visibility detection', () => {
+    it('should refresh right away', () => {
+      let resolver;
+      const promise = new Promise(resolve => {
+        resolver = resolve;
+      });
+      refreshManager.registerElement(
+          testElement, () => {
+            expect(false).to.be.true;
```

Comment:
You can use `fail` here. Also, how can `resolver` ever run? What's the path of execution through this test?